hbase-server/src/main/java/org/apache/hadoop/hbase/master/balancer/HeterogeneousRegionCountCostFunction.java

```diff
@@ -0,0 +1,288 @@
+  private final String rulesPath;
+
+  /**
+   * Contains the rules, key is the regexp for ServerName, value is the limit
+   */
+  private final Map<Pattern, Integer> limitPerRule;
+
+  /**
+   * This is a cache, used to not go through all the limitPerRule map when searching for limit
+   */
+  private final Map<ServerName, Integer> limitPerRS;
+  private int defaultNumberOfRegions;
+
+  /**
+   * Total capacity of regions for the cluster, based on the online RS and their associated rules
+   */
+  private int totalCapacity = 0;
+
+
+  public HeterogeneousRegionCountCostFunction(final Configuration conf) {
+    super(conf);
+    this.limitPerRS = new HashMap<>();
+    this.limitPerRule = new HashMap<>();
+    this.setMultiplier(conf.getFloat(REGION_COUNT_SKEW_COST_KEY, DEFAULT_REGION_COUNT_SKEW_COST));
+
+    this.rulesPath = conf.get(HBASE_MASTER_BALANCER_HETEROGENEOUS_RULES_FILE);
+
+    this.defaultNumberOfRegions = conf.getInt(
+            HBASE_MASTER_BALANCER_HETEROGENEOUS_RULES_DEFAULT, 200);
+    if (this.defaultNumberOfRegions < 0) {
+      LOG.warn("invalid configuration '" + HBASE_MASTER_BALANCER_HETEROGENEOUS_RULES_DEFAULT
+              +"'. Setting default to 200");
+      this.defaultNumberOfRegions = 200;
+    }
+
+    if (conf.getFloat(
+            StochasticLoadBalancer.RegionCountSkewCostFunction.REGION_COUNT_SKEW_COST_KEY,
+            StochasticLoadBalancer.RegionCountSkewCostFunction.DEFAULT_REGION_COUNT_SKEW_COST) > 0)
+    {
+      LOG.warn("regionCountCost is not set to 0, "
+              + " this will interfere with the HeterogeneousRegionCountCostFunction!");
+    }
+  }
+
+  /**
+   * Called once per LB invocation to give the cost function
+   * to initialize it's state, and perform any costly calculation.
+   */
+  @Override
+  void init(final BaseLoadBalancer.Cluster cluster) {
+    this.cluster = cluster;
+    this.loadRules();
+  }
+
+  @Override
+  protected double cost() {
+
+    double cost = 0;
+
+    final double targetUsage = ((double) this.cluster.numRegions / (double) this.totalCapacity);
+
+    for (int i = 0; i < this.cluster.numServers; i++) {
+
+      // retrieve capacity for each RS
+      final ServerName sn = this.cluster.servers[i];
+      final double limit = this.findLimitForRS(sn);
+      final double nbrRegions = this.cluster.regionsPerServer[i].length;
+
+      // cost is the number of regions above the local limit
+      final double usage = nbrRegions / limit;
+      if (usage > targetUsage) {
+        // cost is the number of regions above the local limit
+        final double localCost = (nbrRegions - Math.round(limit * targetUsage)) / limit;
+        cost += localCost;
+      }
+    }
+    return cost / (double) this.cluster.numServers;
+
+  }
+
+  /**
+   * used to load the rule files.
+   */
+  void loadRules() {
+    final List<String> lines = HeterogeneousRegionCountCostFunction.readFile(this.rulesPath);
+    if (lines.size() == 0) {
+      return;
+    }
```

Comment:
Is this cost function still capable of running if no rules are loaded? If not, shouldn't we throw an error to abort it's loading?

Reply:
Yes, default value will be used